File: cecs-424-03-spring-2020-assignment-1/Lab_Assignment_1/C_Sort/sort.c
/*
CECS 424 Spring 2020 Project 1
Name: Randall Thornton
I.D. #016032222
Start Date: 1-21-2020
End Date: 2-11-2020
Description: This program uses a merge sort algorithm 
and a quick sort algorithm to sort any given array of 
size n. Merge sorts a given array by making subarrays of single
vlaues, and then merging these subarrays comparatively.
Quick sorts the given array using a median pivot value
and partitioning elements comparatively.
*/

#include <stdio.h>
#include <stdlib.h>
#include "merge_sort.h"
#include "quick_sort.h"

/* Prints array given passed in length argument */
void print_array(int arr[], int n)
{
    for(int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("\n");    
}
/* Main function establishes requested array from
    assignment and then calls both sorting functions.
    */
int main()
{
    int arr[] = {4, 65, 2, -31, 0, 99, 2, 83, 782, 1};
    int n = 10;
    // int n = sizeof(arr) / sizeof(arr[0]);
    print_array(arr, n);    //Print original array
    qsort2(arr, n);         //Quick sort array
    print_array(arr, n);    //Print sorted array

    printf("\n");

    int arrB[] = {4, 65, 2, -31, 0, 99, 2, 83, 782, 1};
    int nB = 10;
    // int nB = sizeof(arrB) / sizeof(arrB[0]);
    print_array(arrB, nB);  //Print original array
    msort(arrB, nB);        //Quick sort array
    print_array(arrB, nB);  //Print sorted array
}

File: cecs-424-03-spring-2020-assignment-1/Lab_Assignment_1/Haskell_Sort/sort.hs

-- CECS 424 Spring 2020 Project 1
-- Name: Randall Thornton
-- I.D. #016032222
-- Start Date: 1-21-2020
-- End Date: 2-11-2020
-- Description: This is a merge sort
-- and quick sort algorithm that sorts
-- an unsorted array of any given length

-- Declaration of module Sort
module Sort where

merge :: Ord a => [a] -> [a] -> [a] -- merge two sorted lists to a sorted list
merge [] ys = ys                    -- if one
merge xs [] = xs                    --   or the other list is empty merging is trivial
merge (x:xs) (y:ys)                 -- merge two non-empty lists
  | x <= y    = x : merge xs (y:ys)     -- if the left element is smaller then output it first
  | otherwise = y : merge (x:xs) ys     --   otherwise output the right element first

msort :: Ord a => [a] -> [a]    -- Divides halves an unsorted list
msort [] = []                   -- If list is empty return empty
msort [x] = [x]                 -- Return single value since it is a sorted array
msort xs = merge (msort left) (msort right) -- Calls merge of split halves after each half is split again
    where
        left = take (length xs `div` 2) xs  -- Takes first half of the list
        right = drop (length xs `div` 2) xs -- Takes second half of the list

qsort :: Ord a => [a] -> [a]    -- quickly sort an unsorted list given a pivot
qsort [] = []                   -- If array is empty, return empty
qsort [x] = [x]                 -- if array is one value, return it
qsort (pivot : xs) = left ++ [pivot] ++ right   -- Select head as pivot, append left and right with pivot in middle
    where
        left = qsort [y | y <- xs, y <= pivot]  -- sort values of numbers less than pivot
        right = qsort [z | z <- xs, z > pivot]  -- sort values of numbers greater than pivot

main = do 
  let ms = [4,65,2,-31,0,99,2,83,782,1] -- Instantiate arrays
  let qs = [4,65,2,-31,0,99,2,83,782,1]
  print(ms)   -- Print original array
  print(msort ms) -- Print merge sorted array
  print(qs)   -- Original array
  print(qsort qs) -- Quuick sort array